### DISCUSSION TITLE
Proper setup of simulator for linux

### DISCUSSION THREAD

**Discussion by kartun83** [zepp-health/zeppos-forum#165] [Developer Feedback]
Hello.
I'm having trouble setting up simulator in linux mint.

- I followed the instructions from https://docs.zepp.com/docs/guides/tools/simulator/setup/#linux-version-installation
- Simulator without root starts ok and allow to download images and communicate with zeus
- Simulator with root, started with `sudo ./simulator --no-sandbox` throws lots of errors and display just black screen
[simulator.log](https://github.com/zepp-health/zeppos-forum/files/11462945/simulator.log)
- I'm able to separately start  qemu with `sudo /opt/simulator/resources/firmware/start_qemu.sh ZEPP_FW_DIR=/home/kartun/.zepp/emulator_cache/e5d9ef031bef6797a6d10506ca25e493`, but obviously it doesn't have any apps in it
- Launching from simulator window shows only "debug window"
![Screenshot from 2023-05-12 14-09-22](https://github.com/zepp-health/zeppos-forum/assets/5894573/66ef90ea-f823-483e-9fcb-da0211631a86)
![Screenshot from 2023-05-12 14-07-07](https://github.com/zepp-health/zeppos-forum/assets/5894573/815b1e6d-8a7b-4de0-90ed-adb10c525447)

Also it spam with message: `inject stop because application is inconsistent with the watch simulator model`, I checked app.json, and it have definitions for GTR4 and GTS4, and this images are downloaded.
Please advice how to setup this in linux ?

**Comment by kartun83** [comment]
Progress so far
---
For some reason it's listed in docs you could change owner of simulator to yourself, but it's totally not sufficient.
If you check scripts inside /opt/simulator/resources/firmware/qemu_linux both up and down scripts still require root.

One option is to precreate tap0 interface:
`sudo apt-get install uml-utilities && 
sudo nano /etc/network/interfaces`

and put this inside:
```
auto tap0
iface tap0 inet static
address 192.168.166.1
netmask 255.255.255.0
pre-up /usr/bin/tunctl -t tap0 -u <username>
```
 Replace \<username\> with your desired user.
Reboot and check with `ifconfig | grep "tap0"`

----
As you already have "tap0" interface up and running some fixes could be applied to qemu scripts:
open **qemu-ifup** and **qemu-ifdown**, find all strings staring with sudo and comment them, as they are no longer required.

for qemu-ifup it should be like this:
```
#!/bin/sh

echo network interface $1 ip 192.168.166.1 netmask 255.255.255.0
#sudo ifconfig $1 192.168.166.1 netmask 255.255.255.0

echo network interface $1 up
#sudo ifconfig $1 up
```

So, currently you should be able to start qemu without root like this:
`cd /opt/simulator/resources/firmware
./start_qemu.sh ZEPP_FW_DIR=~/.zepp/emulator_cache/<UUID>`

search for correct \<UUID\> inside ~/.zepp/emulator_cache (_assuming you downloaded some images from simulator_)
---
Anyway still doesn't launch qemu from simulator ... :(

**Reply by jiaolaibin** [reply]
For the problem that qemu cannot be launched from simulator:
![image](https://github.com/zepp-health/zeppos-forum/assets/108327182/bd696e8d-a47b-41c6-9f5e-81d431456f86)
Did you not select any products?
You need to click the button in the green box to download the product that needs to be simulated.

**Reply by kartun83** [reply]
imho simulator should display some kind of warning in human readable form instead of stack trace.

**Comment by MyoungXUE** [comment]
Thank you for your feedback, I will feedback this issue to the relevant colleagues.

**Comment by MyoungXUE** [comment]
Has the problem been solved?

**Reply by kartun83** [reply]
partially. Please add parameter to simulator to launch qemu without sudo.
